src/builder.rs

```diff
@@ -0,0 +1,149 @@
+Therefore all errors, related to any of the [`Source`] will only show up then.
+```rust
+# use config::*;
+
+let mut builder = ConfigBuilder::default();
+
+builder.set_default("default", "1").expect("Key is valid");
+builder.add_source(File::new("config/settings", FileFormat::Json));
+builder.set_override("override", "1").expect("Key is valid");
+
+match builder.build() {
+    Ok(config) => {
+        // use your config
+    },
+    Err(e) => {
+        // something went wrong
+    }
+}
+```
+*/
+#[derive(Debug, Clone)]
+pub struct ConfigBuilder {
+    defaults: HashMap<Expression, Value>,
+    overrides: HashMap<Expression, Value>,
+    sources: Vec<Box<dyn Source + Send + Sync>>,
+}
+
+impl ConfigBuilder 
+{
+    /**
+    Creates an empty `ConfigBuilder` without any configuration source registered.
+
+    This is the [`Default`] implementation of `ConfigBuilder`
+    ```rust
+    # use config::ConfigBuilder;
+    let builder = ConfigBuilder::empty();
+    ```
+    */
+    pub fn empty() -> Self {
+        ConfigBuilder {
+            defaults: HashMap::new(),
+            overrides: HashMap::new(),
+            sources: Vec::new(),
+        }
+    }
+
+    /**
+    Set a default `value` at `key`
+    
+    This value can be overwritten by any [`Source`] or override.
+
+    Method can fail if `key` is not valid.
+    */
+    pub fn set_default<T>(&mut self, key: &str, value: T) -> error::Result<&mut ConfigBuilder>
+    where
+        T: Into<Value>,
+    {
+        self.defaults.insert(key.parse()?, value.into());
+        Ok(self)
+    }
+
+    /**
+    Registers new [`Source`] in this builder.
+
+    Calling this method does not invoke any I/O. [`Source`] is only saved in internal register for later use.
+    */
+    pub fn add_source<T>(&mut self, source : T) -> &mut Self 
+    where
+        T: Source + Send + Sync + 'static
+    {
+        self.sources.push(Box::new(source));
+        self
+    }
+
+    /** Set an override
+    
+    This function sets an overwrite value. It will not be altered by any default or [`Source`]
+    
+    Method can fail if `key` is not valid.
+    */
+    pub fn set_override<T>(&mut self, key: &str, value: T) -> error::Result<&mut ConfigBuilder>
+    where
+        T: Into<Value>,
+    {
+        self.overrides.insert(key.parse()?, value.into());
+        Ok(self)
+    }
+
+    /**
+    Reads all registered [`Source`]s.
+
+    This is the method that invokes all I/O operations. 
+    If any of them fail, be it technical reasons or related to inability to read data as `Config` for different reasons,
+    this method returns error.
+    
+    It does not take ownership of `ConfigBuilder` to allow later reuse. Internally it clones data to achieve it.
+    */
+    pub fn build(&self) -> error::Result<Config>  {
+        let mut cache: Value = HashMap::<String, Value>::new().into();
+
+        // Add defaults
+        for (key, val) in self.defaults.iter() {
+            key.set(&mut cache, val.clone());
```

Comment:
Point is that `build` takes `&self`, therefore I cannot use methods that consume struct fields.

It was a conscious decision to do so - as it is written in the comment. I can provide an additional method called for instance `build_owned` that would consume `self`. It would give users the choice - if they do not care about reusing the builder they could just use it. On the other hand, if they do care about it, because they expect configuration to be refreshed, they could stick to `build`. 

Or I could change this method to `build_cloned` and name the one taking ownership `build`?

Reply:
I doubt there is a use-case for not building an owned version, hence I'd use `pub fn build(self) -> ...`.

Also, I'm not sure whether it is possible to have a `fn build(self)` if the previous builder functions only return `&mut Self` instead of `Self` (see #180).

Reply:
I added `build_cloned` and applied your suggestion in `build`.

Reply:
And regarding usefulness - for example in my company we do use a configuration like this in a dynamic way. It makes little sense when all the Sources are files and environment, but as soon as you start adding remote sources be it custom, Consult, or whatever you want, it starts making sense to refresh configuration from time to time in order to take the newest configuration into account.

Reply:
Are you okay with the current state?